### ISSUE TITLE
Cannot build runtime on s390x: The "ApiCompatTask" task failed unexpectedly.

### ISSUE THREAD

**Issue by iii-i** [dotnet/runtime#68266]
### Description

For some weeks it's no longer possible to build the master branch on s390x due to an ApiCompatTask failure.
It seems that this task depends on some DLLs (e.g. `Microsoft.Cci.dll`) that cannot be loaded on s390x.

### Reproduction Steps

On an s390x machine:
- Install dotnet sdk 6.0.
- Run `./build.sh`

### Expected behavior

The build succeeds.

### Actual behavior

The build fails with:

```
/root/.nuget/packages/microsoft.dotnet.apicompat/7.0.0-beta.22179.1/build/Microsoft.DotNet.ApiCompat.targets(52,5): error MSB4018: The "ApiCompatTask" task failed unexpectedly. [/root/dotnet-s390x/runtime/src/libraries/System.Drawing.Primitives/src/System.Drawing.Primitives.csproj]
/root/.nuget/packages/microsoft.dotnet.apicompat/7.0.0-beta.22179.1/build/Microsoft.DotNet.ApiCompat.targets(52,5): error MSB4018: System.TypeLoadException: Could not load type System.Collections.Generic.IEqualityComparer`1[[Microsoft.Cci.ITypeReference, Microsoft.Cci, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a]], System.Runtime, Version=4.2.2.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a while decoding custom attribute: (null) [/root/dotnet-s390x/runtime/src/libraries/System.Drawing.Primitives/src/System.Drawing.Primitives.csproj]
/root/.nuget/packages/microsoft.dotnet.apicompat/7.0.0-beta.22179.1/build/Microsoft.DotNet.ApiCompat.targets(52,5): error MSB4018:    at System.Reflection.CustomAttribute.GetCustomAttributesBase(ICustomAttributeProvider obj, Type attributeType, Boolean inheritedOnly) [/root/dotnet-s390x/runtime/src/libraries/System.Drawing.Primitives/src/System.Drawing.Primitives.csproj]
/root/.nuget/packages/microsoft.dotnet.apicompat/7.0.0-beta.22179.1/build/Microsoft.DotNet.ApiCompat.targets(52,5): error MSB4018:    at System.Reflection.CustomAttribute.GetCustomAttributes(ICustomAttributeProvider obj, Boolean inherit) [/root/dotnet-s390x/runtime/src/libraries/System.Drawing.Primitives/src/System.Drawing.Primitives.csproj]
/root/.nuget/packages/microsoft.dotnet.apicompat/7.0.0-beta.22179.1/build/Microsoft.DotNet.ApiCompat.targets(52,5): error MSB4018:    at System.Attribute.GetCustomAttributes(MemberInfo element, Boolean inherit) [/root/dotnet-s390x/runtime/src/libraries/System.Drawing.Primitives/src/System.Drawing.Primitives.csproj]
/root/.nuget/packages/microsoft.dotnet.apicompat/7.0.0-beta.22179.1/build/Microsoft.DotNet.ApiCompat.targets(52,5): error MSB4018:    at System.Reflection.CustomAttributeExtensions.GetCustomAttributes(MemberInfo element, Boolean inherit) [/root/dotnet-s390x/runtime/src/libraries/System.Drawing.Primitives/src/System.Drawing.Primitives.csproj]
/root/.nuget/packages/microsoft.dotnet.apicompat/7.0.0-beta.22179.1/build/Microsoft.DotNet.ApiCompat.targets(52,5): error MSB4018:    at System.Composition.TypedParts.Util.DirectAttributeContext.GetCustomAttributes(Type reflectedType, MemberInfo member) [/root/dotnet-s390x/runtime/src/libraries/System.Drawing.Primitives/src/System.Drawing.Primitives.csproj]
/root/.nuget/packages/microsoft.dotnet.apicompat/7.0.0-beta.22179.1/build/Microsoft.DotNet.ApiCompat.targets(52,5): error MSB4018:    at System.Composition.Convention.AttributedModelProviderExtensions.GetDeclaredAttributes[ExportAttribute](AttributedModelProvider convention, Type reflectedType, MemberInfo member) [/root/dotnet-s390x/runtime/src/libraries/System.Drawing.Primitives/src/System.Drawing.Primitives.csproj]
/root/.nuget/packages/microsoft.dotnet.apicompat/7.0.0-beta.22179.1/build/Microsoft.DotNet.ApiCompat.targets(52,5): error MSB4018:    at System.Composition.TypedParts.Discovery.TypeInspector.<DiscoverPropertyExports>d__17.MoveNext() [/root/dotnet-s390x/runtime/src/libraries/System.Drawing.Primitives/src/System.Drawing.Primitives.csproj]
/root/.nuget/packages/microsoft.dotnet.apicompat/7.0.0-beta.22179.1/build/Microsoft.DotNet.ApiCompat.targets(52,5): error MSB4018:    at System.Composition.TypedParts.Discovery.TypeInspector.<DiscoverExports>d__0.MoveNext() [/root/dotnet-s390x/runtime/src/libraries/System.Drawing.Primitives/src/System.Drawing.Primitives.csproj]
/root/.nuget/packages/microsoft.dotnet.apicompat/7.0.0-beta.22179.1/build/Microsoft.DotNet.ApiCompat.targets(52,5): error MSB4018:    at System.Composition.TypedParts.Discovery.TypeInspector.InspectTypeForPart(TypeInfo type, DiscoveredPart& part) [/root/dotnet-s390x/runtime/src/libraries/System.Drawing.Primitives/src/System.Drawing.Primitives.csproj]
/root/.nuget/packages/microsoft.dotnet.apicompat/7.0.0-beta.22179.1/build/Microsoft.DotNet.ApiCompat.targets(52,5): error MSB4018:    at System.Composition.TypedParts.TypedPartExportDescriptorProvider..ctor(IEnumerable`1 types, AttributedModelProvider attributeContext) [/root/dotnet-s390x/runtime/src/libraries/System.Drawing.Primitives/src/System.Drawing.Primitives.csproj]
/root/.nuget/packages/microsoft.dotnet.apicompat/7.0.0-beta.22179.1/build/Microsoft.DotNet.ApiCompat.targets(52,5): error MSB4018:    at System.Composition.Hosting.ContainerConfiguration.CreateContainer() [/root/dotnet-s390x/runtime/src/libraries/System.Drawing.Primitives/src/System.Drawing.Primitives.csproj]
/root/.nuget/packages/microsoft.dotnet.apicompat/7.0.0-beta.22179.1/build/Microsoft.DotNet.ApiCompat.targets(52,5): error MSB4018:    at Microsoft.DotNet.ApiCompat.Executor.GetCompositionHost() [/root/dotnet-s390x/runtime/src/libraries/System.Drawing.Primitives/src/System.Drawing.Primitives.csproj]
/root/.nuget/packages/microsoft.dotnet.apicompat/7.0.0-beta.22179.1/build/Microsoft.DotNet.ApiCompat.targets(52,5): error MSB4018:    at Microsoft.DotNet.ApiCompat.Executor.GetDifferenceWriter(TextWriter writer, IDifferenceFilter filter, Boolean enforceOptionalRules, Boolean mdil, Boolean excludeNonBrowsable, Boolean includeInternals, Boolean excludeCompilerGenerated, String remapFile, Boolean groupByAssembly, String leftOperand, String rightOperand, IEnumerable`1 excludeAttributes, Boolean allowDefaultInterfaceMethods, Boolean usesMSBuildLog) [/root/dotnet-s390x/runtime/src/libraries/System.Drawing.Primitives/src/System.Drawing.Primitives.csproj]
/root/.nuget/packages/microsoft.dotnet.apicompat/7.0.0-beta.22179.1/build/Microsoft.DotNet.ApiCompat.targets(52,5): error MSB4018:    at Microsoft.DotNet.ApiCompat.Executor.Run(Boolean usesMSBuildLog, Boolean disableAssemblyResolveTraceListener, IEnumerable`1 contracts, IEnumerable`1 implementationDirectories, TextWriter output, String rightOperand, String leftOperand, Boolean listRules, IEnumerable`1 baselineFileNames, Boolean validateBaseline, Boolean resolveFramework, Boolean skipUnifyToLibPath, IEnumerable`1 contractDependsFileNames, String contractCoreAssembly, Boolean ignoreDesignTimeFacades, Boolean warnOnMissingAssemblies, Boolean respectInternals, Boolean warnOnIncorrectVersion, Boolean enforceOptionalRules, Boolean mdil, Boolean excludeNonBrowsable, Boolean excludeCompilerGenerated, String remapFile, Boolean skipGroupByAssembly, IEnumerable`1 excludeAttributes, Boolean allowDefaultInterfaceMethods) [/root/dotnet-s390x/runtime/src/libraries/System.Drawing.Primitives/src/System.Drawing.Primitives.csproj]
/root/.nuget/packages/microsoft.dotnet.apicompat/7.0.0-beta.22179.1/build/Microsoft.DotNet.ApiCompat.targets(52,5): error MSB4018:    at Microsoft.DotNet.ApiCompat.ApiCompatTask.Execute() [/root/dotnet-s390x/runtime/src/libraries/System.Drawing.Primitives/src/System.Drawing.Primitives.csproj]
/root/.nuget/packages/microsoft.dotnet.apicompat/7.0.0-beta.22179.1/build/Microsoft.DotNet.ApiCompat.targets(52,5): error MSB4018:    at Microsoft.Build.BackEnd.TaskExecutionHost.Microsoft.Build.BackEnd.ITaskExecutionHost.Execute() [/root/dotnet-s390x/runtime/src/libraries/System.Drawing.Primitives/src/System.Drawing.Primitives.csproj]
/root/.nuget/packages/microsoft.dotnet.apicompat/7.0.0-beta.22179.1/build/Microsoft.DotNet.ApiCompat.targets(52,5): error MSB4018:    at Microsoft.Build.BackEnd.TaskBuilder.ExecuteInstantiatedTask(ITaskExecutionHost taskExecutionHost, TaskLoggingContext taskLoggingContext, TaskHost taskHost, ItemBucket bucket, TaskExecutionMode howToExecuteTask) [/root/dotnet-s390x/runtime/src/libraries/System.Drawing.Primitives/src/System.Drawing.Primitives.csproj]
```

### Regression?

Yes. Started with 2575ce5f513c8f5cd37f935ba70b87cbd787c8df.

### Known Workarounds

Cross-building from x86_64 still works.

### Configuration

```
# uname -a
Linux dotnet-s390x 5.4.0-107-generic #121-Ubuntu SMP Thu Mar 24 16:09:01 UTC 2022 s390x s390x s390x GNU/Linux

# dotnet --info
.NET SDK (reflecting any global.json):
 Version:   6.0.100
 Commit:    9e8b04bbff

Runtime Environment:
 OS Name:     ubuntu
 OS Version:  20.04
 OS Platform: Linux
 RID:         linux-s390x
 Base Path:   /root/dotnet-sdk-6.0.100-linux-s390x/sdk/6.0.100/

Host (useful for support):
  Version: 6.0.0
  Commit:  N/A

.NET SDKs installed:
  6.0.100 [/root/dotnet-sdk-6.0.100-linux-s390x/sdk]

.NET runtimes installed:
  Microsoft.AspNetCore.App 6.0.0 [/root/dotnet-sdk-6.0.100-linux-s390x/shared/Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 6.0.0 [/root/dotnet-sdk-6.0.100-linux-s390x/shared/Microsoft.NETCore.App]
```

**Comment by uweigand** [comment]
CC @ViktorHofer

**Comment by jkotas** [comment]
I did not repro the issue locally. 

The key difference between Linux x64 hosted and Linux s390x build is that the Linux x64 hosted build is using CoreCLR and the Linux s390x build is using Mono. If something works fine in Linux x64 hosted and fails in Linux s390x hosted build, it is most likely a Mono bug.

It may be worth it to try building on Linux x64 with CoreCLR overridden with Mono. It should fail as well if my theory is correct.

**Comment by iii-i** [comment]
The x86_64 build with `<PrimaryRuntimeFlavor>Mono</PrimaryRuntimeFlavor>` worked fine for me. This can still be an endianness bug - I'll try to debug further.

**Comment by iii-i** [comment]
After all, this is a Mono vs CoreCLR issue.
The problem with my old test was that I built dotnet with a custom `PrimaryRuntimeFlavor` once, which was not enough (the "host" dotnet did not have that issue, so it did not surface during the build).
Here is the PR with more details in the commit message: https://github.com/dotnet/runtime/pull/70346.

**Comment by lambdageek** [comment]
Fixed by #70346
